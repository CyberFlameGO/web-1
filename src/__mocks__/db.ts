import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';
import { Sequelize } from 'sequelize-typescript';

import { __overrideSequelizeInstanceForTesting } from '../server/db/models';

/**
 * Hooks into jest beforeEach and afterEach to set up and
 * teardown unique database instances usiong sqlite
 */
export const temporaryDatabaseForTestScope = () => {
  let dir: string;

  beforeEach(async () => {
    dir = await fs.promises.mkdtemp(path.resolve(os.tmpdir(), 'cfa-test-db-'));

    await __overrideSequelizeInstanceForTesting(
      new Sequelize({
        dialect: 'sqlite',
        storage: path.resolve(dir, 'test.db'),
        logging: false,
      }),
    );
  });

  afterEach(async () => {
    await fs.remove(dir);
  });
};
