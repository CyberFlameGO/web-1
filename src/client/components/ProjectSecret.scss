.inputContainer {
  display: flex;
  margin-top: 8px;
  margin-bottom: 16px;

  .input {
    flex: 1;
    margin-right: 16px;
  }

  .button {
    user-select: none;
  }
}