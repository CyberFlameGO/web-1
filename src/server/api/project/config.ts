import axios from 'axios';
import * as debug from 'debug';
import * as express from 'express';
import * as Joi from 'joi';

import { validate, hasAdminAccessToTargetRepo } from '../../helpers/_middleware';
import { createA } from '../../helpers/a';
import {
  Project,
  SlackResponderLinker,
  withTransaction,
  CircleCIRequesterConfig,
  TravisCIRequesterConfig,
} from '../../db/models';
import { getProjectFromIdAndCheckPermissions } from './_safe';

const d = debug('cfa:server:api:project:config');
const a = createA(d);

export function configRoutes() {
  const router = express();

  router.post(
    '/:id/config/requesters/circleci',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          accessToken: Joi.string()
            .min(1)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const response = await axios.get(
          `https://circleci.com/api/v1.1/project/github/${project.repoOwner}/${project.repoName}/checkout-key`,
          {
            auth: {
              username: req.body.accessToken,
              password: '',
            },
            validateStatus: () => true,
          },
        );

        if (response.status !== 200) {
          return res.status(401).json({
            error:
              'That token is not valid for the current project, or the repository is not configured on CircleCI',
          });
        }

        const newProject = await withTransaction(async t => {
          const config = await CircleCIRequesterConfig.create({
            accessToken: req.body.accessToken,
          });
          await project.resetAllRequesters(t);
          project.requester_circleCI_id = config.id;
          await project.save({ transaction: t });
          return await Project.findByPk(project.id, {
            include: Project.allIncludes,
            transaction: t,
          });
        });

        res.json(newProject);
      },
    ),
  );

  router.post(
    '/:id/config/requesters/travisci',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          accessToken: Joi.string()
            .min(1)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const response = await axios.get(
          `https://api.travis-ci.org/repo/${project.repoOwner}%2F${project.repoName}`,
          {
            headers: {
              'Travis-API-Version': '3',
              Authorization: `token ${req.body.accessToken}`,
            },
            validateStatus: () => true,
          },
        );

        if (response.status !== 200) {
          return res.status(401).json({
            error:
              'That token is not valid for the current project, or the repository is not configured on Travis CI',
          });
        }

        const newProject = await withTransaction(async t => {
          const config = await TravisCIRequesterConfig.create({
            accessToken: req.body.accessToken,
          });
          await project.resetAllRequesters(t);
          project.requester_travisCI_id = config.id;
          await project.save({ transaction: t });
          return await Project.findByPk(project.id, {
            include: Project.allIncludes,
            transaction: t,
          });
        });

        res.json(newProject);
      },
    ),
  );

  router.post(
    '/:id/config/responders/slack',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        const linker = await withTransaction(async t => {
          await SlackResponderLinker.destroy({
            where: {
              projectId: project.id,
            },
            transaction: t,
          });
          return await SlackResponderLinker.create(
            {
              projectId: project.id,
            },
            {
              transaction: t,
            },
          );
        });

        res.json({
          linker,
          slackClientId: process.env.SLACK_CLIENT_ID,
        });
      },
    ),
  );

  router.patch(
    '/:id/config/responders/slack',
    validate(
      {
        a,
        params: {
          id: Joi.number()
            .integer()
            .required(),
        },
        body: {
          usernameToMention: Joi.string()
            .min(1)
            .max(50)
            .required(),
        },
      },
      async (req, res) => {
        const project = await getProjectFromIdAndCheckPermissions(req.params.id, req, res);
        if (!project) return;

        if (!project.responder_slack) {
          return res.status(400).json({
            error: 'Project is not configured to use Slack as a responder',
          });
        }

        project.responder_slack.usernameToMention = req.body.usernameToMention;
        await project.responder_slack.save();

        res.json(project);
      },
    ),
  );

  return router;
}
