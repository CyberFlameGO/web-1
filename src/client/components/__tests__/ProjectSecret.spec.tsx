/**
 * @jest-environment jsdom
 */

import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { ProjectSecret } from '../ProjectSecret';

import { mockProject } from '../../../__mocks__/project';

describe('<ProjectSecret />', () => {
  it('Should render a disabled input', () => {
    const wrapper = shallow(<ProjectSecret project={mockProject()} />);
    expect(wrapper).toMatchSnapshot();
    const mounted = mount(<ProjectSecret project={mockProject()} />);
    expect(mounted.find('Button')).toHaveLength(1);
  });

  it('should initially dispplay asterix (not the secret)', () => {
    const mounted = mount(<ProjectSecret project={mockProject()} />);
    expect(
      mounted
        .find('Button')
        .children()
        .text(),
    ).toMatchInlineSnapshot(`"Show"`);
    expect(mounted.find('TextInput').prop('value')).toMatchInlineSnapshot(
      `"••••••••••••••••••••••••••••••••••••••••••••••••••"`,
    );
  });

  it('should toggle to display the secret when the button is clicked', () => {
    const mounted = mount(<ProjectSecret project={mockProject()} />);
    mounted.find('Button').simulate('click');
    expect(
      mounted
        .find('Button')
        .children()
        .text(),
    ).toMatchInlineSnapshot(`"Hide"`);
    expect(mounted.find('TextInput').prop('value')).toBe(mockProject().secret);
  });
});
