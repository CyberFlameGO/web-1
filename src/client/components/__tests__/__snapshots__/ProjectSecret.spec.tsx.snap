// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectSecret /> Should render a disabled input 1`] = `
<withTheme(Pane)>
  <withTheme(Heading)>
    Project Secret
  </withTheme(Heading)>
  <withTheme(Pane)
    className="inputContainer"
  >
    <withTheme(TextInput)
      className="input"
      disabled={true}
      value="••••••••••••••••••••••••••••••••••••••••••••••••••"
    />
    <withTheme(Button)
      className="button"
      intent="danger"
      onClick={[Function]}
    >
      Show
    </withTheme(Button)>
  </withTheme(Pane)>
</withTheme(Pane)>
`;
